src/Compiler/Microsoft.NET.Sdk.Razor.SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -200,32 +212,67 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
                     var allTagHelpers = new TagHelperDescriptor[count];
                     tagHelpersFromCompilation.CopyTo(allTagHelpers, 0);
                     tagHelpersFromReferences.CopyTo(allTagHelpers, tagHelpersFromCompilation.Count);
+                    tagHelpersFromComponents.CopyTo(allTagHelpers, tagHelpersFromCompilation.Count + tagHelpersFromReferences.Count);
 
                     return allTagHelpers;
                 });
 
             var generatedOutput = sourceItems
                 .Combine(importFiles.Collect())
-                .Combine(allTagHelpers)
+                .WithLambdaComparer((old, @new) => old.Left.Equals(@new.Left) && old.Right.SequenceEqual(@new.Right), (a) => GetHashCode())
                 .Combine(razorSourceGeneratorOptions)
                 .Select(static (pair, _) =>
                 {
-                    var (((sourceItem, imports), allTagHelpers), razorSourceGeneratorOptions) = pair;
-
-                    RazorSourceGeneratorEventSource.Log.RazorCodeGenerateStart(sourceItem.FilePath);
-
-                    // Add a generated suffix so tools, such as coverlet, consider the file to be generated
-                    var hintName = GetIdentifierFromPath(sourceItem.RelativePhysicalPath) + ".g.cs";
-
-                    var projectEngine = GetGenerationProjectEngine(allTagHelpers, sourceItem, imports, razorSourceGeneratorOptions);
-
-                    var codeDocument = projectEngine.Process(sourceItem);
-                    var csharpDocument = codeDocument.GetCSharpDocument();
-
-                    RazorSourceGeneratorEventSource.Log.RazorCodeGenerateStop(sourceItem.FilePath);
-                    return (hintName, csharpDocument);
-                })
-                .WithLambdaComparer(static (a, b) =>
+                    var ((sourceItem, imports), razorSourceGeneratorOptions) = pair;
+                    
+                    RazorSourceGeneratorEventSource.Log.ParseRazorDocumentStart(sourceItem.RelativePhysicalPath);
+
+                    var projectEngine = GetGenerationProjectEngine(sourceItem, imports, razorSourceGeneratorOptions);
+
+                    var document = projectEngine.ProcessInitialParse(sourceItem);
+
+                    RazorSourceGeneratorEventSource.Log.ParseRazorDocumentStop(sourceItem.RelativePhysicalPath);
+					return (projectEngine, sourceItem.RelativePhysicalPath, document);
+				})
+
+				// Add the tag helpers in, but ignore if they've changed or not, only reprocessing the actual document changed
+				.Combine(allTagHelpers)
```

Comment:
This is the most impactful part of the change:

- We re-write the document if it has changed, but we ignore the status of the tag helpers. 
- The first time through all documents will be changed, so we run with an up-to-date set of tag helpers. 
- Next time through, any documents that have changed will be re-parsed (using the latest tag helpers). Those that haven't changed are skipped (and might have incorrect tag helpers at this point).

Then, on line 253 we re-run not skipping the tag helper check:
- (If neither tag helper not document has changed, there is no work to do)
- When we re-process we pass the `checkIdempotency` flag to the engine. This will see if the same taghelpers that were used last time are being passed this time and skip if so.

This effectively makes the two steps mutually exclusive: we either ran the first step, and thus the tag helpers are up to date and we skip doing anything here; or the document didn't change (so we skipped the first step) but the tag helpers are different meaning we need to re-write them.